### ISSUE TITLE
Concurrent multitab is not working in the chromedp/headless-shell 

### ISSUE THREAD

**Issue by larytet** [chromedp/chromedp#821]
Is it by design? 

Output of the code sample below 
```
2021/05/16 14:28:46 ctx2 iVBORw0KGg
Same browser: false
2021/05/16 14:29:03 websocket url timeout reached

```
```
func main() {
	opts := []ExecAllocatorOption{
		NoFirstRun,
		NoDefaultBrowserCheck,
		NoSandbox,
		// Headless,
		Flag("remote-debugging-port", "9222"),
		Flag("disable-background-networking", true),
		Flag("enable-features", "NetworkService,NetworkServiceInProcess"),
		Flag("disable-background-timer-throttling", true),
		Flag("disable-backgrounding-occluded-windows", true),
		Flag("disable-breakpad", true),
		Flag("disable-client-side-phishing-detection", true),
		Flag("disable-default-apps", true),
		Flag("disable-dev-shm-usage", true),
		Flag("disable-extensions", true),
		Flag("disable-features", "site-per-process,Translate,BlinkGenPropertyTrees"),
		Flag("disable-hang-monitor", true),
		Flag("disable-ipc-flooding-protection", true),
		Flag("disable-popup-blocking", true),
		Flag("disable-prompt-on-repost", true),
		Flag("disable-renderer-backgrounding", true),
		Flag("disable-sync", true),
		Flag("force-color-profile", "srgb"),
		Flag("metrics-recording-only", true),
		Flag("safebrowsing-disable-auto-update", true),
		Flag("enable-automation", true),
		Flag("password-store", "basic"),
		Flag("use-mock-keychain", true),
	}

	// new browser, first tab
	browserCtx, browserCancel := NewExecAllocator(context.Background(), opts...)
	defer browserCancel()

	ctx1, cancel := NewContext(browserCtx)
	defer cancel()

	// ensure the first tab is created
	go func() {
		screenshot := []byte{}
		if err := Run(ctx1, Tasks{
			network.Enable(),
			Navigate("https://www.google.com/search?q=test1"),
			FullScreenshot(&screenshot, 10),
		}); err != nil {
			log.Fatal(err)
		}
		log.Printf("ctx1 %s\n", string(base64.StdEncoding.EncodeToString(screenshot))[:10])
	}()

	// same browser, second tab
	ctx2, _ := NewContext(ctx1)

	// ensure the second tab is created
	go func() {
		screenshot := []byte{}
		if err := Run(ctx2, Tasks{
			network.Enable(),
			Navigate("https://www.google.com/search?q=test2"),
			FullScreenshot(&screenshot, 10),
		}); err != nil {
			log.Fatal(err)
		}
		log.Printf("ctx2 %s\n", string(base64.StdEncoding.EncodeToString(screenshot))[:10])
	}()

	time.Sleep(20 * time.Second)

	c1 := FromContext(ctx1)
	c2 := FromContext(ctx2)

	fmt.Printf("Same browser: %t\n", c1.Browser == c2.Browser)
	fmt.Printf("Same tab: %v %v %t\n", c1.Target, c2.Target, c1.Target == c2.Target)

	// Output:
	// Same browser: true
	// Same tab: false
}

```

**Comment by larytet** [comment]
I can not make multiple tabs/browsers work in parallel with the chromedp/headless-shell

**Comment by ZekeLu** [comment]
Here is the copy of my answer from another issue. Let's keep the discussion here.

> I have took a peak at your code, my suggestion is do not make things complicated at the very first. Make simple thing work first, then extend it with more features.
> 
> > `Flag("remote-debugging-port", "9222")`
> 
> I guess this is why it does not work. If you use `chromedp.NewExecAllocator` to start a browser, you don't need to specify the port. Let the browser choose one which is not in use for you.
> 
> It seems that this is your first time to use `chromedp`. Please read through the document (https://pkg.go.dev/github.com/chromedp/chromedp) and the examples (https://github.com/chromedp/examples) first. Then try something simple to understand how `chromedp` works.
> 
> And here is the suggestion for using `golang`:
> 
> 1. don't use **dot imports** (https://github.com/golang/go/issues/29326)
> 2. if you start from scratch, don't use `go 1.13`. The last two supported versions are `1.15` and `1.16`.

**Comment by larytet** [comment]
remote-debugging-port was the culprit

**Comment by larytet** [comment]
In case anyone reaches this page while looking for "What is the best practice when using NewExecAllocator".

At the moment I call NewExecAllocator() once - one context, create 4-20 tabs (NewContext) depending on the available resources. I never cancel any one of the contexts unless the tab fails. I have a pool of free tabs (a pool of available tab contexts) and reuse it. The implementation is in https://github.com/larytet-go/headlessness

**Comment by ZekeLu** [comment]
@larytet No, you're creating several browser instances instead of several tabs. You can see it with the following simple demo:

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"net/http/httptest"
	"time"

	"github.com/chromedp/chromedp"
)

func main() {
	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Println("request:", r.URL)
		if r.URL.Path == "/index" {
			fmt.Fprintf(w, `
<html>
<body>
hello
</body>
</html>`)
		}
	}))
	defer s.Close()

	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("headless", false),
	)
	aCtx, cancel := chromedp.NewExecAllocator(context.Background(), opts...)
	defer cancel()

	ctx1, cancel := chromedp.NewContext(aCtx,
		chromedp.WithDebugf(log.Printf),
	)
	defer cancel()

	if err := chromedp.Run(ctx1,
		chromedp.Navigate(s.URL+"/index"),
	); err != nil {
		log.Println(err)
	}

	ctx2, cancel := chromedp.NewContext(aCtx,
		chromedp.WithDebugf(log.Printf),
	)
	defer cancel()

	if err := chromedp.Run(ctx2,
		chromedp.Navigate(s.URL+"/index"),
	); err != nil {
		log.Println(err)
	}

	time.Sleep(20 * time.Second)
}
```

Check the example to see how to create several tabs in a browser:

https://github.com/chromedp/chromedp/blob/d3ef975bdb93712fc152b4719300e19a84c6769f/example_test.go#L200-L227

And regarding reusing tabs, I think it's better to create a new tab for a new task, and close when the job is done. Here are the reasons:
1. If a tab is waiting in the pool, its resources are not released;
2. It should be better to start from a clean state (a new tab);
3. How to replace a broken tab (I'm not sure whether there will be one) in the pool?
4. open/close a tab in an existing browser is cheap.

**Comment by larytet** [comment]
I refactored the code. The flow is NewExecAllocator->NewContext - I keep the result, but I never use it directly. Instead I call NewContext() ... cancel() every time I need a tab. I do not see a measurable difference in the startup times. I guess any approach works just fine.
